src/audio/module_adapter/module_adapter.c

```diff
@@ -858,7 +858,7 @@ static int module_adapter_audio_stream_copy_1to1(struct comp_dev *dev)
 		buffer_stream_writeback(sink_c, mod->output_buffers[0].size);
 
 	if (mod->output_buffers[0].size)
-		audio_stream_produce(&sink_c->stream, mod->output_buffers[0].size);
+		comp_update_buffer_produce(sink_c, mod->output_buffers[0].size);
```

Comment:
@serhiy-katsyuba-intel @btian1 @ranj063 IIUC, this code was changed (in reverse to this patch) to improve MCPS. I think the comp_ API was calling some notifiers (that I think were only used for legacy WOV and L1 exit and no longer used). I would say here we need a light touch when invoking a probe or injector. We could remove the notifier and directly check for probe/injector.

Reply:
I did some performance measurements in comp_update_buffer_produce(). When CONFIG_SOF_LOG_LEVEL_DBG is enabled, debug logs at the bottom of that func consumes o lot of CPU. Entire function consumes 11233 ccounts of which two buf_dbg() consume 10419 ccounts. Removing buf_dbg() reducess CPU consumtion more then 10 times.

Without buf_dbg() and when no listeners registered, comp_update_buffer_produce() seems does not do anything which consumes lot's of CPU. If needed, it could be potentially optimized more: there is a loop there over all cores to check if IDC is needed to send notification to another core -- that loop could be eliminated with prior "if" as usually (always?) notification sender and receiver is on same core. But such optimization seems will not give us much.

I modified "Enable buffer produce notifications required by probe" commit by commenting out those expensive buf_dbg() calles. Hope now it's OK to return comp_update_buffer_produce() back to module_adapter.

High CPU consumption only happens if CONFIG_SOF_LOG_LEVEL_DBG is enabled. Do you recall if DBG level was enabled when performance measurements were made previously?